okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java

```diff
@@ -69,21 +74,27 @@ public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
     spec.apply(sslSocket, getOkHttpRoute(host, port, spec));
 
     Platform platform = Platform.get();
-    try {
-      // Force handshake.
-      sslSocket.startHandshake();
-
-      String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
-      if (negotiatedProtocol == null) {
-        throw new RuntimeException("protocol negotiation failed");
+    // It's possible that the user provided SSLSocketFactory has already done the handshake
```

Comment:
The only reason I don't want getSession() is that it swallows the exception if there is an error during the handshake, although it returns an invalid session object with an invalid cipher suite of "SSL_NULL_WITH_NULL_NULL" so that we can detect the failure, but we lost the error message and stack trace which should be useful to tell users what's going on.